funcx_endpoint/funcx_endpoint/endpoint/interchange.py

```diff
@@ -321,53 +321,85 @@ def _main_loop(self):
-                    log.debug(f"Submitting task : {task}")
-                    executor.submit_raw(task)
-                except queue.Empty:
-                    pass
-                except Exception:
-                    log.exception("Unhandled issue while waiting for pending tasks")
-
-                try:
-                    results = self.results_passthrough.get(False, 0.01)
-
-                    task_id = results["task_id"]
-                    if task_id:
-                        self.results_ack_handler.put(task_id, results["message"])
-                        log.info(f"Passing result to forwarder for task {task_id}")
-
-                    message = try_convert_to_messagepack(results["message"])
-
-                    # results will be a packed EPStatusReport or a packed Result
-                    log.debug(f"Publishing message {message}")
-                    results_publisher.publish(message)
-                    self.results_ack_handler.ack(task_id)  # no error; RMQ has it now
-                except queue.Empty:
-                    pass
-
-                except Exception:
-                    log.exception(
-                        "Something broke while forwarding results from executor "
-                        "to forwarder queues"
-                    )
-                    continue
+
+            num_tasks_forwarded = 0
+            num_results_forwarded = 0
+
+            def process_pending_tasks():
+                nonlocal num_tasks_forwarded
+                while not self._quiesce_event.is_set():
+                    try:
+                        incoming_task = self.pending_task_queue.get(timeout=1)
+                        task = try_convert_from_messagepack(incoming_task)
+                        executor.submit_raw(task)
+                        num_tasks_forwarded += 1  # Safe given GIL
+
+                    except queue.Empty:
+                        continue
+
+                    except Exception:
+                        log.exception("Unhandled issue while waiting for pending tasks")
+
+                log.debug("Exit process-pending-tasks thread.")
+
+            def process_pending_results():
+                nonlocal num_results_forwarded
+                while not self._quiesce_event.is_set():
+                    try:
+                        results = self.results_passthrough.get(timeout=1)
+
+                        task_id = results["task_id"]
+                        if task_id:
+                            self.results_ack_handler.put(task_id, results["message"])
+                            log.debug(f"Forwarding result for task {task_id}")
+
+                        message = try_convert_to_messagepack(results["message"])
+
+                        # results will be a packed EPStatusReport or a packed Result
+                        log.debug(f"Publishing message {message}")
+                        results_publisher.publish(message)
```

Comment:
As discussed, the plan is to have `publish` in a separate try block, to handle the case where it fails say due to a disconnect. We can do this in tandem with #869